File: src/main/scala/generics/scala/Fruit.scala
package generics.scala

trait Fruit {
  def color(): String
}

trait Eatable extends Fruit {
  def joice(): String
}

trait Toxic extends Fruit {
  def detoxify(): Eatable
}

case class Orange() extends Eatable() {
  override def joice(): String = "Orange joice"
  val color = "Orange"
}

class Apple(val color: String) extends Eatable() {
  override def joice(): String = "Apple joice"

  def removeAppleCores(): String = "Apple cores removed"
}

class ProcessedAckee(val color: String) extends Eatable() {
  override def joice(): String = "Ackee joice"
}

class Ackee extends Toxic {
  override def color(): String = "Red"

  override def detoxify(): ProcessedAckee = new ProcessedAckee("Orange");
}

File: src/main/scala/generics/java/covariance/Fruit.java
package generics.java.covariance;

interface Fruit {
	String color();
};

class Orange implements Fruit {
	public String color() {
		return "Orange";
	}
}

class Apple implements Fruit {
	private final String color;

	public Apple(String color) {
		this.color = color;
	}

	@Override
	public String color() {
		return color;
	}

	public String joice() {
		return "Apple joice";
	}
}


